salt/modules/x509_v2.py

```diff
@@ -0,0 +1,1933 @@
+        - x509.sign_remote_certificate
+
+In order for the :term:`Compound Matcher` to work with restricting signing
+policies to a subset of minions, in addition calls to :py:func:`match.compound <salt.modules.match.compound>`
+by the minion acting as the CA must be permitted:
+
+.. code-block:: yaml
+
+    # /etc/salt/master.d/peer.conf
+
+    peer:
+      .*:
+        - x509.sign_remote_certificate
+
+      ca_server:
+        - match.compound
+
+.. note::
+
+    Compound matching in signing policies currently has security tradeoffs since the
+    CA server queries the requesting minion itself if it matches, not the Salt master.
+    It is recommended to rely on glob matching only.
+
+Signing policies
+~~~~~~~~~~~~~~~~
+In addition, the minion representing the CA needs to have at least one
+signing policy configured, remote calls not referencing one are always
+rejected.
+
+The parameters specified in this signing policy override any
+parameters passed from the minion requesting the certificate. It can be
+configured in the CA minion's pillar, which takes precedence, or any
+location :py:func:`config.get <salt.modules.config.get>` looks up in.
+Signing policies are defined under ``x509_signing_policies``.
+
+You can restrict which minions can request a certificate under a configured
+signing policy by specifying a matcher in ``minions``. This can be a glob
+or compound matcher (for the latter, see the notes above).
+
+.. code-block:: yaml
+
+    x509_signing_policies:
+      www:
+        - minions: 'www*'
+        - signing_private_key: /etc/pki/ca.key
+        - signing_cert: /etc/pki/ca.crt
+        - C: US
+        - ST: Utah
+        - L: Salt Lake City
+        - basicConstraints: "critical, CA:false"
+        - keyUsage: "critical, cRLSign, keyCertSign"
+        - subjectKeyIdentifier: hash
+        - authorityKeyIdentifier: keyid,issuer:always
+        - days_valid: 90
+        - copypath: /etc/pki/issued_certs/
+
+
+.. note::
+
+    The following semantics are applied regarding order of preference
+    for specifying the subject name:
+
+    * If neither ``subject`` nor any name attributes (like ``CN``) are part of the policy,
+      issued certificates can contain any requested ones.
+    * If any name attributes are specified in the signing policy, ``subject`` contained
+      in requests is ignored.
+    * If ``subject`` is specified in the signing policy, any name attributes are ignored.
+      If the request contains the same data type for ``subject`` as the signing policy
+      (for dicts and lists, and only then), merging is performed, otherwise ``subject``
+      is taken from the signing policy. Dicts are merged and list items are appended,
+      with the items taken from the signing policy having priority.
+
+Breaking changes versus the previous ``x509`` modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+General
+^^^^^^^
+* ``path`` and ``text`` parameters have been removed for all ``create_*`` functions.
```

Comment:
Perhaps, support and issue deprecation warnings?

Reply:
It really depends. Initially, I wanted this to be a drop-in replacement like the Vault stuff, but since this module needs to handle the serialization/hydration of all extensions (previously, M2Crypto/OpenSSL handled that part), there is always going to be some breakage without applying a substantial amount of effort. Hence I decided to start with a clean slate.

I can definitely try to reduce the list as much as possible though. It was primarily intended as a reference for the reviewers and myself to know what the differences are, not necessary the final list users should have to look at.

easy fix:
* renamed parameters
* changed defaults
* removed parameters can be accepted with a warning that they are ignored (e.g. `version` cannot be set with `cryptography`, same for `verbose`.)
* removing irrelevant stuff from the list (eg `sign_remote_certificate` is not something users have to worry about. The implication is that minions using the old module are not compatible with a CA server using the new ones)

medium effort:
* support writing to files in execution module

major effort:
* ensuring the output/configuration format for extensions is the same

Reply:
Okay, I reduced the breaking changes a lot. The transition should be quite smooth now for anyone using the basic functionality. There will be a lot of deprecation warnings though.

The remaining breaking changes are:
* output change/possible extension configuration change (The latter likely only for more exotic configurations. Not sure how some of the more complicated extensions were done in the old module as it was never actually documented. I doubt many actually used more than the basic ones, see `issuserAltName` typo).
* RDNs in the subject are ordered the other way around (which is the industry standard), causing changes during the first run only
* `x509.private_key_managed` defaulting mode to `0400`, which is a bug fix imho
* CA server and minion still need to be on the same version of the module

Mind that I still need to write the tests for the deprecation and writing to files stuff, but wanted to push the changes as early as possible.